### PR TITLE
Fix race around thread_create/thread_launch/thread_destroy

### PR DESCRIPTION
**Pull request by jakubkarolczyk**

### CODE REVIEW

File: src/include/libks/ks_threadmutex.h [hunk 1]
```diff
@@ -83,6 +82,11 @@ KS_DECLARE(ks_status_t) __ks_thread_create_ex(ks_thread_t **thread, ks_thread_fu
 #define ks_thread_create(thread, func, data, pool)						\
 	__ks_thread_create_ex(thread, func, data, KS_THREAD_FLAG_DEFAULT, KS_THREAD_DEFAULT_STACK, KS_PRI_NORMAL, pool, __FILE__, __LINE__,  __PRETTY_FUNCTION__)
 
+KS_DECLARE(ks_status_t) __ks_thread_destroy_ex(ks_thread_t **threadp, ks_bool_t internal_call);
```
Comment (andywolk) [inline src/include/libks/ks_threadmutex.h:85]: What does `internal_call` mean for the end-user?

File: src/ks_thread.c [hunk 2]
```diff
@@ -389,7 +413,12 @@ static ks_status_t __init_os_thread(ks_thread_t *thread)
 	}
 #endif
 
+	ks_mutex_lock(thread->mutex);
+	thread->inuse = KS_TRUE;
+
 	if ((err = pthread_create(&thread->handle, &thread->attribute, thread_launch, thread)) != 0) {
+		thread->inuse = KS_FALSE;
+		ks_mutex_unlock(thread->mutex);
```
Comment (andywolk) [inline src/ks_thread.c:421]: This `in_use = false` and the unlock will make `thread` structure be possibly freed and therefore un-toucheable but it's used later in the code still.

**Review by andywolk — CHANGES_REQUESTED** [review]

File: src/include/libks/internal/ks_thread.h [hunk 3]
```diff
@@ -44,4 +44,7 @@ struct ks_thread {
 	uint8_t priority;
 	void *return_data;
 	ks_pool_t *pool_to_destroy;
+
+	ks_mutex_t *mutex;
+	ks_bool_t inuse;
```
[the remark this answers is not in the document]
Comment (jakubkarolczyk) [reply src/include/libks/internal/ks_thread.h:49]: np, I'll change it

File: src/include/libks/ks_threadmutex.h (hunk 1 above)
Comment (jakubkarolczyk) [reply src/include/libks/ks_threadmutex.h:85]: internal_call TRUE can be passed to __ks_thread_destroy_ex only within ks_thread.c. __ks_thread_destroy_ex shouldn't be used outside this context. End user and also other parts of libks should use ks_thread_destroy, which by default pass FALSE as internal_call

Comment (andywolk) [reply src/include/libks/ks_threadmutex.h:85]: `KS_DECLARE` tells me it's a part of the API and I can use it.

Comment (andywolk) [reply src/include/libks/ks_threadmutex.h:85]: Don't give users a chance then. Make `__ks_thread_destroy_ex` static.

File: src/ks_thread.c (hunk 2 above)
Comment (jakubkarolczyk) [reply src/ks_thread.c:421]: In the situation when end user runs ks_thread_create and ks_thread_destroy asynchronously, yes. If this is something possible, I can fix, np.

File: src/include/libks/ks_threadmutex.h (hunk 1 above)
Comment (jakubkarolczyk) [reply src/include/libks/ks_threadmutex.h:85]: ok, np

File: src/ks_thread.c (hunk 2 above)
Comment (jakubkarolczyk) [reply src/ks_thread.c:421]: Adding this I actually give the end user possiblity to call ks_thread_destroy async, which doesn't make sense, I'll change it

Comment (andywolk) [reply src/ks_thread.c:421]: Well. Looked again and I don't think that user knows the thread to call destroy at that point. So async is not possible.
I should take my words back.

File: src/ks_thread.c [hunk 4]
```diff
@@ -389,7 +413,12 @@ static ks_status_t __init_os_thread(ks_thread_t *thread)
 	}
 #endif
 
+	ks_mutex_lock(thread->mutex);
```
Comment (andywolk) [inline src/ks_thread.c:416]: What are you protecting against here with that lock?

Comment (jakubkarolczyk) [reply src/ks_thread.c:416]: Discussed verbaly

**Review by andywolk — APPROVED** [review]

**Pull request merged**